examples/tool/src/middlewares/headers.middleware.ts

```diff
@@ -0,0 +1,12 @@
+import type { Get, UniversalMiddleware } from "universal-middleware";
+
+const headersMiddleware: Get<[], UniversalMiddleware> =
+  () => (_request, ctx) => {
+    return (response) => {
+      response.headers.set("X-Custom-Header", ctx.something ?? "NONE");
```

Comment:
I didn't review everything, but I'll be honest, stuff like this rubs me all the wrong ways - requiring optional fields in the context type, and then using the `??` operator, it just sounds too brittle for my taste. Users will put those middlewares in the wrong order, things will quietly malfunction - because the types aren't really telling the truth, it will be difficult for them to understand why.

I really want to see type safety for this.

Reply:
Type safety can be achieved with a proper `webroute-adapter` for instance.
Proper type safety cannot be achieved currently with the current state of javascript servers (without something like `webroute`), which is out of scope here.

The scope of this PR is to ease the work of tool authors, and the whole middleware composition stack should not be a part of this solution, as it adds a lot of overhead, maintainability cost, and complexity when explaining what this project purpose is.

It's still a huge step forward in the right direction. And again, complete type safety will be possible with `webroute` and the servers it supports.

Reply:
> It's still a huge step forward in the right direction.

Yeah, no doubt. 🙂

> And again, complete type safety will be possible with `webroute` and the servers it supports.

How?

If this universal middleware has inaccurate context types, how would `webroute` solve that?

Reply:
I'm currently working on some improvements and the `webroute` adapter to make it clearer.

Reply:
[Example using webroute adapter](https://github.com/magne4000/universal-middleware/blob/1c923ac3327cb11905a70d0388c829e1647f86c1/packages/adapter-webroute/tests/entry-webroute.ts)

cc @mindplay-dk @sinclairnick

The middleware signature has been adapted to correspond [this comment](https://github.com/magne4000/universal-middleware/discussions/2#discussioncomment-10365051).

Reply:
Apologies if I've misunderstood here, as there's obviously a lot of context/info in this repo I may have missed.

But if I understand correctly, there is indeed a type that's allows declaring the inputs and outputs, without having to rely on inference.

The [MiddlewareFn](https://github.com/sinclairnick/webroute/blob/feature/docs-rewrite/packages/middleware/src/types.ts) type enables setting:
- The "data result"
- The middleware's extra input parameters
- The "response handler"'s extra parameters

```ts
export type MiddlewareFn<
  T extends DataResult | void = void,
  TRest extends any[] = any[],
  TRestRes extends any[] = any[]
> = (
  request: Request,
  ...rest: TRest
) => Awaitable<MiddlewareResult<T, TRestRes>>;
```

So to achieve what you're describing this would look something like:

```ts
type MyMiddleware = MiddlewareFn<OutputCtx, [InputCtx], []>
// OutputCtx is the return type, InputCtx is a single extra param, no response handler extra params
```

Obviously a small wrapper type could bend this into the correct order of params etc.

Reply:
@sinclairnick thanks, that's a good start, but I realize that I should make a reproduction repo. My issue seems to be more related to the interaction between `MiddlewareFn` from `@webroute/middleware` used by `route().use(...)`.